### PR TITLE
Heartbeat Agency Endpoint Update

### PR DESCRIPTION
**Pull request by maierlars**
The heartbeat thread updates the agency endpoint once per second but did not put the leader in front, instead it put them in lexicographic order. Thus in case the leader is not the first server, each server in the cluster forgets every second which server was the actual leader and asked the first named follower, just to be redirected immediately.

### CODE REVIEW

File: arangod/Cluster/HeartbeatThread.cpp [hunk 1]
```diff
@@ -1306,8 +1306,15 @@ void HeartbeatThread::updateAgentPool(VPackSlice const& agentPool) {
       agentPool.hasKey("size") && agentPool.get("size").getUInt() > 0) {
     try {
       std::vector<std::string> values;
+      // we have to make sure that the leader is on the front
+      auto leaderId = agentPool.get("id").stringRef();
+      auto pool = agentPool.get("pool");
+      values.emplace_back(pool.get(leaderId).copyString());
+      // now add all non leaders
       for (auto pair : VPackObjectIterator(agentPool.get("pool"))) {
```
Comment (jsteemann) [inline arangod/Cluster/HeartbeatThread.cpp]:
```suggestion
      for (auto pair : VPackObjectIterator(pool)) {
```

File: arangod/Cluster/HeartbeatThread.cpp [hunk 2]
```diff
@@ -1306,8 +1306,15 @@ void HeartbeatThread::updateAgentPool(VPackSlice const& agentPool) {
       agentPool.hasKey("size") && agentPool.get("size").getUInt() > 0) {
     try {
       std::vector<std::string> values;
+      // we have to make sure that the leader is on the front
+      auto leaderId = agentPool.get("id").stringRef();
+      auto pool = agentPool.get("pool");
+      values.emplace_back(pool.get(leaderId).copyString());
```
Comment (jsteemann) [inline arangod/Cluster/HeartbeatThread.cpp]: We can also add a `values.reserve(pool.length());` here to save some reallocations later.

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/7146/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/7175/

**Comment by neunhoef** [comment]
This can and should be merged to 3.5 for 3.5.4

**Pull request merged**